sql/2021/resource-hints/dns-prefetch_domains.sql

```diff
@@ -0,0 +1,65 @@
+#standardSQL
+# Most popular domains users dns-prefetch from
+
+# helper to create percent fields
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+  ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+CREATE TEMPORARY FUNCTION getResourceHints(payload STRING, hint STRING)
+RETURNS STRING
+LANGUAGE js AS '''
+try {
+  var $ = JSON.parse(payload);
+  var almanac = JSON.parse($._almanac);
+  return JSON.stringify(almanac['link-nodes'].nodes.filter(link => link.rel.toLowerCase() == hint));
+} catch (e) {  
+  return "";
+}
+''' ;
+
+CREATE TEMPORARY FUNCTION getHrefs(payload STRING)
+RETURNS ARRAY<STRING>
+LANGUAGE js AS '''
+try {
+    var $ = JSON.parse(payload);
+    return $.map(n => n.href);
+} catch (e) {
+    return [];
+}
+''' ;
```

Comment:
The SQL takes the output of the first function and passes it into the second function, so they could be combined.
```suggestion
CREATE TEMPORARY FUNCTION getResourceHintHrefs(payload STRING, hint STRING)
RETURNS ARRAY<STRING>
LANGUAGE js AS '''
try {
  var $ = JSON.parse(payload);
  var almanac = JSON.parse($._almanac);
  return almanac['link-nodes'].nodes.filter(link => link.rel.toLowerCase() == hint).map(link => link.href);
} catch (e) {
  return [];
}
''';
```